### PR TITLE
Typedoc library mode api docs

### PR DESCRIPTION
**Pull request by telamonian**
## References

fixes: #8598

Builds the API docs using the [new experimental "library" mode](https://github.com/TypeStrong/typedoc/pull/1184) in typedoc. This allows for a nice monorepo-wide landing page, as well as for live links between our `@jupyterlab/whatever` packages in the docs.

image

image

image

### CODE REVIEW

File: tsconfigdoc.json [hunk 1]
```diff
@@ -0,0 +1,254 @@
+{
+  "$schema": "http://json.schemastore.org/tsconfig",
+  "extends": "./tsconfigbase",
+  "compilerOptions": {
+    "composite": false,
+    "baseUrl": ".",
+    "paths": { "@jupyterlab/*": ["packages/*/src"] }
+  },
+  "references": [
```
Comment (blink1073) [inline tsconfigdoc.json:9]: We should generate this in `buildutils/src/ensure-repo.ts`

File: typedoc.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
+// const git = require('git-rev-sync');
+
+const packages = [
+  'application-extension',
```
Comment (blink1073) [inline typedoc.js:4]: This should also be generated.

Comment (telamonian) [reply typedoc.js:4]: So far as I can tell, `inputFiles` always gets ignored, so the `packages` array here probably isn't needed. I'm still experimenting

File: tsconfigdoc.json (hunk 1 above)
Comment (telamonian) [reply tsconfigdoc.json:9]: I agree

**Comment by telamonian** [comment]
As of the latest commit to this PR:

- All core jlab packages now have `strictNullChecks` enabled
- All core jlab packages are now included in the API docs build

Though the API docs in this PR have a few rough edges, they are already vastly superior to our existing API docs. I think we should pull this PR in and publish new API docs post-haste.

Remaining problems:

- The packages are listed as eg `"application-extension/src/index"` instead of just `application-extension` in the docs TOC
- The README.md files for individual packages are not currently included in the API docs (though the top level README.md is)
- (probably not an actual problem) The default exports of a package (which we only use for extension plugin objects in core) are not included in the docs

I don't think any of these should be considered blocking.

pinging @blink1073 @afshin for a second opinion

File: tsconfigdoc.json (hunk 1 above)
Comment (telamonian) [reply tsconfigdoc.json:11]: autogen enabled

File: typedoc.js (hunk 2 above)
Comment (telamonian) [reply typedoc.js]: I removed that array entirely

**Pull request merged**